### PR TITLE
update features docs

### PR DESCRIPTION
**Pull request by L1aoXingyu**

### CODE REVIEW

**Comment by L1aoXingyu** [comment]
image
image
image
image

File: docs/source/tutorials/Features.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+# Features in LiBai
+
+LiBai provides many features out of the box. We will show you how easily configuring them step by step.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Features.md:3]: how to？

File: docs/source/tutorials/Features.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+## Automatic Mixed Precision Training
+
+AMP stands for automatic mixed precision training. In LiBai, we allowed the user to train with AMP with just configuration. You can just simply add `amp` in your configuration file to use AMP.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Features.md:7]: allow用一般时，表示过去，现在，将来都可以，而不仅仅是过去允许？

File: docs/source/tutorials/Features.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+# enable amp
+train.amp.enabled = True
+# disable amp
+train.amp.enabled = False
+```
+
+## Gradient Clipping
+
+Gradient clipping is a technique that tackles exploding gradients. The idea of gradient clipping is very simple: If the gradient gets too large, we rescale it to keep it small.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Features.md:27]: 冒号后面的if是小写?

File: docs/source/tutorials/Features.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+optim.params.clip_grad_norm_type = None
+```
+
+`clip_grad_max_norm` and `clip_grad_norm_type` can be checked in [API docs of oneflow](https://oneflow.readthedocs.io/en/master/nn.html#oneflow.nn.utils.clip_grad_norm_).
+
+## Gradient Accumulation
+
+Gradient accumulation is a common strategy to enlarge your batch size for training. When training with large-scale models, memory can easily become the bottleneck. But decreasing the batch size (e.g., 2) will lead to unsatisfactory convergence.
+
+Besides, when training with pipeline parallel, gradient accumulation make different stages executed in parallel micro-batch. Therefore, the calculation of each stage is not blocked.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Features.md:58]: gradient accumulation makes

File: docs/source/tutorials/Features.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+# enable grad accumulation for 4 steps
+train.num_accumulation_steps = 4
+
+# disable grad accumulation
+train.num_accumulation_steps = None
+```
+
+## Activation Checkpointing
+
+To reduce GPU memory usage and deploy a large model to a training system, LiBai support activation checkpointing. We use a Transformer layer as the unit of checkpointing because the activation size bloats in the middle of a Transformer layer so checkpointing the input of a Transformer layer is storage-efficient.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Features.md:79]: LiBai supports?

File: docs/source/tutorials/Features.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+LiBai supported activation checkpointing by `set_activation_checkpoint` in `GraphBase`. So models using `libai.layers.TransformerLayer` supported activation checkpointing by default. If you want to set activation checkpointing for your customized layers, you need to override this function.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Features.md:81]: 应该是一直支持，不用过去时

File: docs/source/tutorials/Features.md (hunk 2 above)
Comment (Ldpe2G) [inline docs/source/tutorials/Features.md:7]:
```suggestion
AMP stands for automatic mixed precision training. In LiBai, to enable AMP you can simply add `train.amp.enabled = True` to your configuration file.
```

Comment (Ldpe2G) [reply docs/source/tutorials/Features.md:7]: 感觉这样简洁一些

File: docs/source/tutorials/Features.md (hunk 3 above)
Comment (Ldpe2G) [inline docs/source/tutorials/Features.md:27]:
```suggestion
Gradient clipping is a technique that tackles exploding gradients. The idea of gradient clipping is very simple: the gradient will be rescaled down if it gets too large.
```

Comment (Ldpe2G) [reply docs/source/tutorials/Features.md:27]: 这样如何

File: docs/source/tutorials/Features.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+train.amp.enabled = True
+# disable amp
+train.amp.enabled = False
+```
+
+## Gradient Clipping
+
+Gradient clipping is a technique that tackles exploding gradients. The idea of gradient clipping is very simple: If the gradient gets too large, we rescale it to keep it small.
+
+You do not have to worry about implementing gradient clipping when using LiBai, we support gradient clipping in a convenient way. You just need to add it in your configuration file.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Features.md:29]:
```suggestion
LiBai support gradient clipping in a convenient way, you don't have to implement it by yourself. You just need to add some setting to your configuration file to enable gradient clipping.
```

File: docs/source/tutorials/Features.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+# disable gradient clipping
+optim.params.clip_grad_max_norm = None
+optim.params.clip_grad_norm_type = None
+```
+
+`clip_grad_max_norm` and `clip_grad_norm_type` can be checked in [API docs of oneflow](https://oneflow.readthedocs.io/en/master/nn.html#oneflow.nn.utils.clip_grad_norm_).
+
+## Gradient Accumulation
+
+Gradient accumulation is a common strategy to enlarge your batch size for training. When training with large-scale models, memory can easily become the bottleneck. But decreasing the batch size (e.g., 2) will lead to unsatisfactory convergence.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Features.md:56]:
```suggestion
Gradient accumulation is a common strategy to train large-scale models when memory becomes the bottleneck. This technique will split the original batch to several micro batches, then perform normal forward and backward operation, the model will only be updated after accumulating the gradients of all these micro batches.
```

Comment (Ldpe2G) [reply docs/source/tutorials/Features.md:56]: 感觉可以解释的更详细一些

File: docs/source/tutorials/Features.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+# enable activation checkpointing
+train.activation_checkpoint.enabled = True
+
+# disable activation checkpointing
+train.activation_checkpoint.enabled = False
+```
+
+## ZeRO 
+
+Unlike basic data parallelism where memory states are replicated across data-parallel processes, Zero Redundancy Optimizer (ZeRO) partitions model states and gradients to save significant memory.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Features.md:109]:
```suggestion
Unlike normal data parallelism, where model states and gradients are replicated across data-parallel processes. Zero Redundancy Optimizer (ZeRO) partitions them across data-parallel process which can reduce memory footprint significantly.
```

File: docs/source/tutorials/Features.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+- Level 1: The optimizer states (e.g., for Adam optimizer, 32-bit weights, and the first, and second moment estimates) are partitioned across the processes, so that each process updates only its partition.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Features.md:111]:
```suggestion
- Level 1: The optimizer states (e.g., for Adam optimizer, 32-bit weights, and the first, and second moment estimates) are partitioned across the processes, so that each process wil only update its own partition.
```

File: docs/source/tutorials/Features.md (hunk 2 above)
Comment (rentainhe) [reply docs/source/tutorials/Features.md:7]: 我也感觉这样表达更合适~~

File: docs/source/tutorials/Features.md (hunk 5 above)
Comment (rentainhe) [reply docs/source/tutorials/Features.md:79]: Activation Checkpointing是不是要加个paper或者啥的链接比较好 LiBai support [activation checkpointing]()

File: docs/source/tutorials/Features.md (hunk 3 above)
Comment (L1aoXingyu) [reply docs/source/tutorials/Features.md:27]: 冒号后面如果是句子，需要大写第一个字母。

File: docs/source/tutorials/Features.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+optim.params.clip_grad_norm_type = None
+```
+
+`clip_grad_max_norm` and `clip_grad_norm_type` can be checked in [API docs of oneflow](https://oneflow.readthedocs.io/en/master/nn.html#oneflow.nn.utils.clip_grad_norm_).
+
+## Gradient Accumulation
+
+Gradient accumulation is a common strategy to train large-scale models when memory becomes the bottleneck. This technique will split the mini-batch into several micro-batches, then perform normal forward and backward operations. Models will only be updated after accumulating the gradients of all these micro-batches. 
+
+Besides, when training with pipeline parallel, gradient accumulation makes different stages executed in parallel micro-batch. Therefore, the calculation of each stage is not blocked.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Features.md:58]:
```suggestion
Besides, when training with pipeline parallel, gradient accumulation makes different stages executed different micro-batch in parallel. Therefore, the calculation of each stage can be overlapped.
```

Comment (Ldpe2G) [reply docs/source/tutorials/Features.md:58]: 这样如何

**Review by Ldpe2G — APPROVED** [review]
